### PR TITLE
JSON (and CBOR/MessagePack) Tabular and Timeseries Outputs

### PR DESCRIPTION
**Pull request by mbadams5**
Pull request overview
---------------------
This PR is dependent on #5893 going in first since this is built off that branch. The following link provides a more accurate comparison of the features in this specific PR, [seen here.](https://github.com/NREL/EnergyPlus/compare/input_processor_refactor...output_processor_json)

This PR adds JSON outputs for Tabular and Timeseries outputs. Due to the JSON library we are using for the JSON input processing, we can also output [CBOR](http://cbor.io/) and [MessagePack](http://msgpack.org/) binary formats. Both of these formats provide faster serialization/deserialization and much smaller file sizes.

The JSON output format was built off of work completed a couple years ago and updated to current EnergyPlus source code and JSON library.

Running the Outpatient Reference Model for an annual simulation has the following file sizes:

eplusout.mtr - 2.1M
eplusmtr.csv - 1.6M
eplusout.eso - 31M
eplusout.csv - 24M
eplusout.sql - 31M

eplusout_hourly.json - 42M
eplusout_hourly.cbor - 12M (~71% reduction to JSON, 50% compared to CSV)

eplusout.json - 5.0M
eplusout.cbor - 1.2M (76% reduction)

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [ ] Code style (parentheses padding, variable names)
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check -- verify this
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add rules to spreadsheet
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Documentation changes in place
 - [ ] Changed docs build successfully
 - [ ] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

### CODE REVIEW

**Comment by nrel-bot-3** [comment]
@mbadams5 @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@mbadams5 @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@mbadams5 @lgentile it has been 15 days since this pull request was last updated.

**Comment by Myoldmopar** [comment]
@mbadams5 so you said this branch is built on top of the input json branch, so the input json branch should go in first.  But this one has idd changes, so we're trying to squeeze it in first actually.  I'm assuming once this is clean we'll just merge this one in so the other one can go away.

On this branch, here's what I see as I start prepping to review:
image

Could you advise on whether this is really something that thing be fixed up and ready to go in by Monday morning?  I know it's important to try to get this in, so the door is being held open a few days past IDD freeze.  I'm sure the conflicts, while there are a bunch, are easy to resolve.  The failing builds are what concerns me.

If the IDD changes aren't going to be ready, there can still be other things going in after IDD freeze, of course.

**Comment by JasonGlazer** [comment]
@mbadams5 I am working on the output units refactoring #6157 and wanted to check with you on possible overlap of code. Are you changing anything related to SetupOutputVariables? Or related to any of the functions listed [here ](https://github.com/NREL/EnergyPlus/blob/OutputUnitsRefactor/design/FY2017/DesDoc-OutputUnitsRefactor.md#approach)? Your diffs are not all showing on GitHub, probably because there are so many files changed.  If there are overlaps, I would like to understand them so that merging goes as easily as possible.

**Comment by nrel-bot-3** [comment]
@mbadams5 @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@mbadams5 @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@mbadams5 @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@mbadams5 @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@mbadams5 @lgentile it has been 15 days since this pull request was last updated.

**Comment by Myoldmopar** [comment]
@mbadams5 is there a reason you didn't wrap the json input file [parse](https://github.com/NREL/EnergyPlus/blob/15e7d45cf5225d928dcd1b21f7e295cf5c464464/src/EnergyPlus/InputProcessing/InputProcessor.cc#L284) call with a try block?  Is there a try block at the next level up in the callstack?

**Comment by nrel-bot-3** [comment]
@mbadams5 @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@mbadams5 @lgentile it has been 14 days since this pull request was last updated.

**Comment by Myoldmopar** [comment]
So, this won't be making it in now, right?  I'm marking it for the next milestone to get it off of my review list.

**Comment by mbadams5** [comment]
@Myoldmopar I had planned to get this in. I've been working through build errors due to Reporting Frequency and Units changes, but i'm nearly done with that.

**Comment by Myoldmopar** [comment]
@mbadams5 I was hoping to find this all green when I woke, but there are too many issues here to try to drop in this morning.  Build warnings, license text issues, diffs, encoding problems when Python is trying to process output files, and maybe others.  This won't make it.  But at least the input JSON got in!

**Comment by nrel-bot-2** [comment]
@mbadams5 @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@mbadams5 @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@mbadams5 @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@mbadams5 @lgentile it has been 14 days since this pull request was last updated.

**Comment by Myoldmopar** [comment]
@mbadams5 I mean, there are some issues here and there, but I feel like this isn't really too far from ready.  What are your feelings about wrapping it up this week?

**Comment by Myoldmopar** [comment]
@mbadams5 I'm going to go ahead and mark this for the next release.  If something crazy happens and you are ready for this to go in real soon then just let me know and feel free to change the milestone back.

**Comment by Myoldmopar** [comment]
Looks like some license text issues, pedantic build warnings, one JSON output file failing, and Clang found a unit test failure (I think the expected value just needs to be updated, but I'd say verify with the history of that unit test file.)  Actually, the JSON output file succeeds it looks like, but maybe it's not getting a proper exit code for some reason or something else is going on.  Not sure.  But it looks like this is indeed very close now!

**Comment by jasondegraw** [comment]
@mbadams5 I was looking through the modifications, and I was wondering: does this support yearly outputs?

**Comment by mbadams5** [comment]
@jasondegraw It should... Is there a specific reason why you think it doesn't?

**Comment by jasondegraw** [comment]
@mbadams5 I'm seeing separate files for different frequencies but not one for yearly.

File: idd/Energy+.idd.in [hunk 1]
```diff
@@ -106702,6 +106702,31 @@ Meter:CustomDecrement,
    A46, \field Key Name 22
    A47; \field Output Variable or Meter Name 22
 
+Output:JSON,
+       \memo Output from EnergyPlus can be written to JSON format files.
+       \unique-object
+  A1 , \field Option Type
+  	   \required-field
```
Comment (mjwitte) [inline idd/Energy+.idd.in]: There's a stray tab here.

File: idd/Energy+.idd.in [hunk 2]
```diff
@@ -106702,6 +106702,31 @@ Meter:CustomDecrement,
...
+       \type choice
+       \key TimeSeries
+       \key TimeSeriesAndTabular
```
Comment (mjwitte) [inline idd/Energy+.idd.in]: Do we want a default here?

**Review by mjwitte — COMMENTED** [review]
Are there doc changes for this?

**Comment by jasondegraw** [comment]
@mbadams5 @mjwitte One other thing, maybe it's minor but I'll bring it up anyway. Right now there are two output frequencies that mean essentially the same thing: RunPeriod and Environment. Both get mapped into the frequency enum value `Simulation`, so there's a third name for essentially the same thing. I was thinking we should unify on "Environment" as the one name and get rid of the others, and that would require renaming all the `*_runperiod.*` filenames to `*_environment.*`. Any thoughts?

File: scripts/runenergyplus.in [hunk 3]
```diff
@@ -263,6 +263,32 @@ rm -f  eplusout.log
 rm -f  eplusout.svg
 rm -f  eplusout.shd
 rm -f  eplusout.wrl
+
+rm -f  eplusout.json
+rm -f  eplusout_detailed_zone.json
+rm -f  eplusout_detailed_HVAC.json
+rm -f  eplusout_timestep.json
+rm -f  eplusout_hourly.json
```
Comment (mjwitte) [inline scripts/runenergyplus.in:271]: @jasondegraw I'm ok with dropping RunPeriod. Do you want to make that transition for this release? So, I'm surmising from this question that these new outputs get sorted by reporting frequency into separate files? Looking at these names, we don't need all three of these: detailed_zone, detailed_HVAC, and timestep. detailed_zone and timestep are the same frequency.

Comment (jasondegraw) [reply scripts/runenergyplus.in:271]: @mjwitte Sooner is better than later, I'd be fine with just renaming the file(s?) for now, then cleaning up the internal stuff (change the enum value to `Environment`) and the IDD later.

Comment (mjwitte) [reply scripts/runenergyplus.in:271]: Let's do that as a separate pull request so it doesn't bog down this one. Simple idd change with transition and doc change and modifying any testfiles, of course that may need it, and . . .

**Comment by Myoldmopar** [comment]
@mbadams5 I haven't seen any potential encoding-fix-commits coming in, so I'm going to push this to the next release, as we try to freeze for IO this afternoon.

**Comment by nrel-bot-3** [comment]
@mbadams5 @lgentile it has been 28 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@mbadams5 @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@mbadams5 it has been 8 days since this pull request was last updated.

**Comment by mbadams5** [comment]
@Myoldmopar @mjwitte I think this is nearly done. The encoding fixes I did on the inputs side were needed on the outputs side as well, as I have mentioned previously. One remaining issue that is solvable, but I don't particularly want to do it in this PR, is when the weather stats file has a different encoding than UTF-8.

Many (or all) of the included stats files in the weather directory are `Western (Windows 1252)` encoded, which causes issues when searching and outputting to UTF-8. I think I've fixed the code so that it will not fail with errors when it is encoded with Windows 1252, however, it does not output 100% correctly.

HTML output with `Western (Windows 1252)` encoded stats file
Windows-1252

HTML output with `UTF-8` encoded stats file
UTF-8

Note the bad encoding of the degree symbol for latitude and longitude. Also all the lines related to Köppen classification and ASHRAE Handbook CDD/HDD are missing with the `Western (Windows 1252)` encoded stats file.

It can't find the Köppen classification because it is [searching the `Western (Windows 1252)` line using an `UTF-8` encoded string](https://github.com/NREL/EnergyPlus/blob/a7c9cc14ce05cf09dca6623de97faa8546e2e1bb/src/EnergyPlus/OutputReportTabular.cc#L5646). We'd need to either re-write the specific parts to be encoding agnostic, read the file stats file as UTF-8, or convert the encoding of that specific input string to UTF-8.

**Comment by mjwitte** [comment]
For this particular table, it seems worthwhile to work around it.  Would it work to search just parts of the string, leaving out the offending characters, and then output the full utf-8 string, not what's in the stat file? In fact, I'd prefer to see the line show up in the table with a blank value even if the data isn't found in the stat file.

**Comment by mbadams5** [comment]
@mjwitte I get that, but this isn't specific to this PR. I just compiled develop which has the stat file at `Western (Windows 1252)` and OutputReportTabular.cc at `Western (Windows 1252)`. When I change either the stat file or OutputReportTabular.cc to UTF-8 encoding, I get the following results:

Stat: `Western (Windows 1252)`
OutputReportTabular.cc: `Western (Windows 1252)`
Western (Windows 1252)/Western (Windows 1252)

Stat: `UTF-8`
OutputReportTabular.cc: `Western (Windows 1252)`
UTF-8/Western (Windows 1252)

Stat: `UTF-8`
OutputReportTabular.cc: `UTF-8`
UTF-8/UTF-8


Stat: `Western (Windows 1252)`
OutputReportTabular.cc: `UTF-8`
Western (Windows 1252)/UTF-8

**Comment by mjwitte** [comment]
@mbadams5 Sure, it can be a separate issue to make this table more robust, just do the right thing here for now and post a followup.

**Comment by Myoldmopar** [comment]
@mbadams5 I pulled this locally, merged in latest develop, and am going to test it out.  I am in the process of pointing E+ CI to the updated regression scripts, so that UTF-8 issue should go away very soon.  Other than some final testing, are you ready for this to go in?  @mjwitte anything else on your end?

**Comment by mjwitte** [comment]
@Myoldmopar The only part of this I've looked at/commented on is the encoding issue related to the Climatic Data Summary noted [above](https://github.com/NREL/EnergyPlus/pull/6192#issuecomment-439106724). I've posted #7102 as a followup issue to fix this.

**Comment by Myoldmopar** [comment]
I just ran the UTF8 test file on the new regression scripts with this branch and develop and it passes nicely.  Which means that once CI is totally using the new scripts (later this week - maybe tomorrow), that error will go away.  Nothing to be concerned about with this branch.  Otherwise this branch seems OK.  I'll do a final pass over the code and confirm it on the call in a few minutes before merging.

File: scripts/runenergyplus.in (hunk 3 above)
Comment (Myoldmopar) [reply scripts/runenergyplus.in:271]: @mjwitte any change of heart on this since August?

Comment (mjwitte) [reply scripts/runenergyplus.in:271]: At this point, do whatever you see fit. No strong opinion on this.

**Comment by Myoldmopar** [comment]
@mbadams5 I tried this out this morning.  I pulled latest develop in and re-built.  I'll put some findings here as I go.

- I added an Output:JSON, TimeSeries; object to 1ZoneUncontrolled.idf
   - Expected: Output files for each time series frequency
   - Outcome: No JSON output files
- I then added a "Yes" to field A2 on the Output:JSON object, and re-ran, and got the outputs!  This seems like a bug, as field A2 should be defaulted to Yes according to the IDD.

Spot checked some results and everything looks good and ran well.  I think this is ready to drop in if you can address that issue.  I'm still perusing the code one final time right now but it's good so far.

**Comment by mbadams5** [comment]
@Myoldmopar I just pushed up a fix. See if that works.

**Comment by Myoldmopar** [comment]
Pulled, built, and tested.  JSON now outputs properly even if nothing is entered for field A2.  I also reported cbor outputs, and imported them in python using a [cbor](https://pypi.org/project/cbor/) package.  It looks good when reported back out as normal JSON.  I am happy with this.

I will let CI have a pass over it this afternoon and if it looks good, I'll merge it in.  Thanks @mbadams5 !

**Comment by Myoldmopar** [comment]
All green!  Merging.

**Comment by DavidHollman** [comment]
> ... The following link provides a more accurate comparison of the features in this specific PR, [seen here.](https://github.com/NREL/EnergyPlus/compare/input_processor_refactor...output_processor_json)
> 

The link goes to a page that says "There isn’t anything to compare."

Is there an updated way to do that?

**Pull request merged**